Kadira Debug Support for Production Apps

kasunben opened the pull request.

kasunben committed:
Inits

kasunben committed:
check for appId

kasunben committed:
single session allowing for production apps

arunoda commented on lib/client/connect.js:
```diff
@@ -6,7 +6,11 @@ Meteor.subscribe('kadira.debug.listeners');
 KadiraDebug._init = function(browserId, clientId) {
-  Meteor.subscribe('kadira.debug.init', browserId, clientId);
+
+  var debug = getQueryParamFromURI("debug");
```
May be set `kadira_debug` rather than debug.

arunoda commented on lib/client/store.js:
```diff
@@ -243,4 +243,11 @@ Store.prototype._runCallbacks = function(callbacks) {
\ No newline at end of file
+StoreManager = window.StoreManager = new Store();
+
+// function getQueryParamFromURI(q) {
```
Remove these.

arunoda commented on lib/server/connect.js:
```diff
@@ -1,3 +1,10 @@
+AppConfig = {
+  env: getAppEnv(),
+  authKey: getDebugAuthKey(),
+  accessToken: [],
```
We need to keep accessTokens inside a LRU Cache.

arunoda commented on lib/server/connect.js:
```diff
@@ -10,14 +17,25 @@ SubHandlers = {
+        return;
+      }
+
+      AppConfig.sessionsAuthInfo[sessionId] = true;
```
Keep this like this.

`KadiraInfo.session.__kadiraDebugAuthorized = true`

arunoda commented on lib/server/connect.js:
```diff
@@ -69,6 +100,27 @@ Meteor.methods({
+      appEnv = AppConfig.env;
+
+      accessToken = generateAccessToken();
+      AppConfig.accessToken.push(accessToken);
```
Put this into a LRU cache.

arunoda commented on lib/server/trace_store.js:
```diff
@@ -2,6 +2,7 @@ var LRUCache = Npm.require('lru-cache');
 
 TraceStore = function() {
   this._sessionMapper = {};
+  this.registeredSessions = [];
```
Use a map.

arunoda commented on lib/server/trace_store.js:
```diff
@@ -107,6 +117,12 @@ TraceStore.prototype._trackTime = function(type, sessionId, id, event, timestamp
 };
 
 TraceStore.prototype._onMethodTrace = function(trace, session) {
+  if(!_.contains(this.registeredSessions, session.id)) {
```
Use `registeredSessions` as a map and use a code like below:

``` js
if(!this.registeredSessions[session.id]) {
  return;
}
```

arunoda commented:
We also need a method to clean all the sessions. Then Kadira Debug UI can invoke it to clean sessions.

kasunben committed:
improve some code logic

kasunben commented:
Updated.
*\* except `cleanAllSession` method

arunoda commented on lib/server/connect.js:
```diff
@@ -69,6 +103,28 @@ Meteor.methods({
     return traceStore.getTrace(browserId, clientId, type, id);
+  },
+
+  'kadira.debug.getAccessToken': function(authKey) {
```
Change the name into `kadira.debug.createAccessToken`

arunoda commented on lib/server/connect.js:
```diff
@@ -100,6 +156,35 @@ Meteor.publish('kadira.debug.listeners', function() {
+  return 'production';
+}
+
+function getDebugAuthKey() {
```
Change the name as `getAuthKey`

arunoda commented on lib/server/trace_store.js:
```diff
@@ -17,11 +18,18 @@ TraceStore = function() {
   delete this._sessionMapper[key];
+
+  // unRegister the session
+  this.registeredSessions[sessionId] = false;
```
Delete the the field. Otherwise, we'll collecting a lot of sessionIds.

```
delete this.registeredSessions[sessionId];
```

arunoda commented on lib/server/trace_store.js:
```diff
@@ -107,6 +115,13 @@ TraceStore.prototype._trackTime = function(type, sessionId, id, event, timestamp
 
 TraceStore.prototype._onMethodTrace = function(trace, session) {
+  var validSession = this.registeredSessions[session.id] || false;
+  if(!validSession) {
```
Simply check it in the if. 

```
if(this.registeredSessions[session.id]) {
   ..
}
```

arunoda commented on lib/server/trace_store.js:
```diff
@@ -119,6 +134,13 @@ TraceStore.prototype._onMethodTrace = function(trace, session) {
 };
 
 TraceStore.prototype._onSubTrace = function(trace, session) {
+  var validSession = this.registeredSessions[session.id] || false;
```
Same as above.

arunoda commented on lib/server/connect.js:
```diff
@@ -10,14 +19,26 @@ SubHandlers = {
+        return;
+      }
+
+      AppConfig._authorizedSessions.set(sessionId, true); // ::
```
Don't use another LRU cache for this. Mark the session as a authorized one. Like this:

```
kadiraInfo.session.__kadiraDebugSessionAuthrozied = true;
```

kasunben replied:
here;
1. `kadiraInfo.session` can't use for store any other data because it already use for keep `sessionId` as a single value.

arunoda replied:
Ah got it. Then assign it to `kadiraInfo` directly.

kasunben committed:
remove temporaly added authKey and dev values

kasunben committed:
add kadira.debug.reset() method

kasunben committed:
add test cases for createAccessToken

kasunben committed:
authKey logic changed

kasunben committed:
add kadira:runtime-dev as a dependancy and use it for detect the env.

kasunben committed:
KADIRA_DEBUG_AUTH_KEY authentication added.

kasunben committed:
remove test authKey

kasunben committed:
revised createToken test case

kasunben committed:
add Session Authorization for createAccessToken method + change test case according to that

kasunben commented:
added Session Authorization for `createAccessToken` method + change test case according to that

arunoda commented on lib/server/connect.js:
```diff
@@ -69,10 +104,54 @@ Meteor.methods({
+  'kadira.debug.reset': function() {
+    AppConfig.accessToken.reset();
+    AppConfig._authorizedClientSessions = {};
+    AppConfig._authorizedServerSessions = {};
```
Change this as: `_authorizedAdminSessions`

arunoda commented on lib/server/connect.js:
```diff
@@ -69,10 +104,54 @@ Meteor.methods({
+    return AppConfig.env;
+  },
+
+  'kadira.debug.createAccessToken': function(authKey) {
```
No need to pass authKey here.

arunoda commented on lib/server/connect.js:
```diff
@@ -69,10 +104,54 @@ Meteor.methods({
+    // when invalid auth key
+    // return without proceed
+    if((AppConfig._authorizedServerSessions[sessionId]) 
+      && (authKey !== AppConfig.authKey)) {
```
Why  this?

kasunben committed:
code improvements

arunoda commented on lib/client/connect.js:
```diff
@@ -45,4 +49,11 @@ if(!browserId) {
   KadiraDebug._init(browserId, clientId);
+}
+
+function getQueryParamFromURI(q) {
```
Can't we use a npm module for that? Try to find a small npm module for this if it's there. Make sure, how much of code it's add to the app.

arunoda commented on lib/server/connect.js:
```diff
@@ -10,17 +20,29 @@ SubHandlers = {
+      if(!AppConfig.accessToken.has(accessToken)) {
+        // If the subcription request with an invalid accessToken
+        // return without subscribe
+        return;
```
You should not return a publication like this. That doesn't stop the publication. Do these:
- Throw a Meteor.Error() (I like that approach)
- Do `return this.stop()`;
- Do `return []` (I don't like this)

arunoda commented on lib/server/connect.js:
```diff
@@ -32,6 +54,18 @@ Meteor.methods({
+      if(!AppConfig._authorizedClientSessions[sessionId]) {
+        // not a valid session
+        // return without proceed
+        return;
```
Same as the previous comment.
See: https://github.com/kadirahq/meteor-debug/pull/20/files#r43823648

arunoda commented on lib/server/connect.js:
```diff
@@ -69,10 +103,52 @@ Meteor.methods({
+    // return without proceed
+    if(!AppConfig._authorizedAdminSessions[sessionId]) {
+      throw new Meteor.Error("Unauthorized", "Invalid AuthKey or Session!.");
+      return;
```
No need to return after throwing an Error.
Did you lint this code?

arunoda commented on lib/server/connect.js:
```diff
@@ -69,10 +103,52 @@ Meteor.methods({
+      return;
+    }
+
+    var accessToken = generateToken(4);
```
Use `Random.id()`
You need to depend on the `random` package. It's a must.
I assume there is an option to specify the limit of the string.

arunoda commented on lib/server/connect.js:
```diff
@@ -69,10 +103,52 @@ Meteor.methods({
+    }
+
+    var accessToken = generateToken(4);
+    AppConfig.accessToken.set(accessToken, sessionId);
```
Change the name into `AppConfig.accessTokens`

arunoda commented on lib/server/connect.js:
```diff
@@ -69,10 +103,52 @@ Meteor.methods({
+    } else {
+      // return when invalid authKey
+      throw new Meteor.Error("Unauthorized");
+      return;
```
No need to return.

arunoda commented on lib/server/connect.js:
```diff
@@ -69,10 +103,52 @@ Meteor.methods({
+      AppConfig._authorizedAdminSessions[sessionId] = true;
+    } else {
+      // return when invalid authKey
+      throw new Meteor.Error("Unauthorized");
```
In `Meteor.Error`. First argument is a number. Like this:

```
throw new Meteor.Error("401", "Unauthorized")
```

arunoda commented on lib/server/connect.js:
```diff
@@ -69,10 +103,52 @@ Meteor.methods({
+    // when invalid auth key
+    // return without proceed
+    if(!AppConfig._authorizedAdminSessions[sessionId]) {
+      throw new Meteor.Error("Unauthorized", "Invalid AuthKey or Session!.");
```
Use "401" instead of "Unauthorized"

arunoda commented on lib/server/connect.js:
```diff
@@ -87,12 +163,14 @@ Meteor.publish('kadira.debug.timeline', function() {
 Meteor.publish('kadira.debug.listeners', function() {
+  var kadiraInfo = Kadira._getInfo();
+  var sessionId = kadiraInfo.session;
+
```
You need to authorize here as well. May be it's better to move the auth code to a common place.

arunoda commented on lib/server/connect.js:
```diff
@@ -100,6 +178,29 @@ Meteor.publish('kadira.debug.listeners', function() {
   });
 });
 
+function getAppEnv() {
```
Move all these to a new file called `utils.js` and create a new global namespace `Utils`. And assign these functions to it.

arunoda commented:
Some core changes:
- Prefix all the method/pubsub names like this:
  - `kadira.debug.admin` - If called from the Kadira Debug UI
  - `kadira.debug.client` - If called from the the App's client
- Create two functions to authorize for admin and client
  - If development no check allow access
  - If session is marked as authenticated, allow access
  - If not throw an Meteor.Error("401")
- Use it on top of the all the method/pubsub calls
- For the authentication we do it via a new publication. All the authentication code goes into that. We have two of them for both client and admin like:
  - `kadira.debug.admin.auth`
  - `kadira.debug.client.auth`
- These publication lives throughout the lifecycle of the connection.

arunoda commented:
Add these test cases. 
- [ ] Client - Test queryparams getting code
- [x] Authentication based on access token
  - [x] success
  - [x] failed
  - [x] on development
  - [x] delete the code when the publication stops
- [ ] Authorize based on the access token
  - [ ] if the session authorize
  - [ ] if not
  - [ ] on development
- [x] Access Token Management
  - [x] create it
  - [ ] reset it
- [x] Authentication based on authKey
  - [x] if key does not match
  - [x] if key matches
  - [x] delete the code when the publication stops
- [x] Authorize based on the auth key
  - [x] if the session authorize
  - [x] if not
  - [x] on development
- [x] Trace Store
  - [x] do not track method traces coming from not registered sessions
  - [x] do not track pubsub traces coming from not registered sessions

kasunben committed:
Error handling and change code arrangement

kasunben committed:
fix minor issue

kasunben committed:
change method/pubsub prefix

kasunben committed:
add authentication pub/sub

kasunben commented:
Updated with;
- prefix changes
- authorised function
- authentication pub/sub

kasunben committed:
rearrange pub/sub and method order

kasunben committed:
Test Cases

kasunben commented:
Test cases added.

arunoda commented on lib/server/connect.js:
```diff
@@ -10,7 +22,92 @@ SubHandlers = {
+});
+
+Meteor.publish('kadira.debug.remote.timeline', function() {
+
```
remove this space.

arunoda replied:
Check whether this connection(or publication) is stopped, when reconnecting from the Kadira Debug UI.

arunoda commented on lib/server/connect.js:
```diff
@@ -10,7 +22,92 @@ SubHandlers = {
+
+  var kadiraInfo = Kadira._getInfo();
+
+  var isAuthorized = isAuthorizedSession('remote', kadiraInfo);
```
Can't we get kadiraInfo right inside `isAuthorizedSession` ?

arunoda commented on lib/server/connect.js:
```diff
@@ -10,7 +22,92 @@ SubHandlers = {
+  var kadiraInfo = Kadira._getInfo();
+
+  var isAuthorized = isAuthorizedSession('remote', kadiraInfo);
+  if(!isAuthorized) {
```
You don't need to check it here. 
If not authorized, isAuthorizedSession will throw the error.

arunoda commented on lib/server/connect.js:
```diff
@@ -105,4 +214,13 @@ function updateListenersCount() {
   });
+}
+
+function isAuthorizedSession(type, kadiraInfo) {
```
Prefix this as `KadiraDebug._isAuthorizedSession`.

Then we can write unit tests for this.

kasunben committed:
add KadiraDebug._authorize() for isAuthorizedSession() method

kasunben committed:
test cases revised

kasunben committed:
add sessionId arg to KadiraDebug._authorize()

kasunben committed:
delete the code when the publication stops

kasunben commented:
Test cases updated.
:
- “delete the code when the publication stops” tests checking in “Authentication based on access token - success’  and “Authentication based on authKey - Success”
- Access Token Management  - reset it : no different method for reset,

thinkholic commented:
- [ ] set pre-loader for traces and Debug URL generation
- [ ] add LRU cache for traces
- [ ] add `copy to clip board` button for debug URL
- [ ] `bug` Once reset, debug session not starts accordingly (sometimes)
- [ ] add Full screen Icon and 
- [ ] remove “We love your feedback link”
- [ ] add `Create Debug Session` button instead of Link
- [ ] move `reset` button to right

arunoda committed:
Reviewed connect.js

arunoda committed:
Reviewed server/connect.js

arunoda committed:
Reviewed server side code

kasunben committed:
Better connection logic in the client

kasunben committed:
Implementing better start functionality for client side store

kasunben committed:
fix some minor issue

kasunben committed:
add _authorize method to few other server method calls

kasunben committed:
revert store.js changes due timeline issue

kasunben committed:
fix trace unavailable on prod. issue

kasunben committed:
fix and modified isStated logic on StoreManager

kasunben committed:
Run  only if  subscription is ready

kasunben committed:
Update Test Cases

kasunben committed:
fix some error handlings on client/connect.js and Store._isStarted()

kasunben committed:
fix Test case failing on tests/client/store.js

kasunben committed:
authorize getTrace method

kasunben committed:
fix 'remove timeline sub handle after disconnected' integation test issue

kasunben committed:
add _sleepForMs for 'remove timeline sub handle after disconnected' test case

arunoda committed:
Fix integration tests a bit

arunoda committed:
Merge pull request #25 from kadirahq/kd-production-review

kd-production-review - Fixed & Updated

arunoda committed:
Get debugAuthKey from Meteor.settings too.
Now it try to get it from Meteor.settings, if there is
no env. variable of 'KADIRA_DEBUG_AUTH_KEY'.
It get the authKey from 'Meteor.settings.kadira.debug.authKey'

arunoda committed:
Merge branch 'kd-production-review' into kd-production
